utils/README.md

```diff
@@ -20,6 +25,77 @@ enough to deserve their own module.
+few extension functions to allow dynamic string interpolation with any object.
+
+### String syntax
+
+A string that can be interpolated is a string that contains variables, which are delimited by `{}`. The name of the
+variable is the content of the braces and must be an alphanumeric string identifier that can contain underscores and
+dashes. `"Hello {name}, I'm {my-name} and I'm {my_age} years old."` is a valid string that contains 3 variables.
+
+Any opened brace must be closed, and opening braces can be escaped using a backslash (`\{`). For example, the string
+`"Hello {name"` is invalid but `"Hello \{name"` is valid because the opening brace is escaped.
+
+Failure to respect these rules will result in an exception being thrown.
+
+### Interpolation
+
+To interpolate a string, one must use the `interpolate` extension function on a `String`. It exists in several variants.
+
+The first ones are `interpolate(vararg args: Pair<String, Any>)` and `interpolate(args: Map<String, Any>)`. They take a
+list of pairs of variable names associated with their values or a map with the same semantics. The variable names must
+be valid identifiers, and the values can be of any type. The function will replace all variables in the string with
+their associated value.
+
+```kt
+val string = "Hello I'm {name}, and I'm {age} years old.".interpolate(
+    "name" to "John",
+    "age" to 42,
+)
+print(string) // prints "Hello I'm John, and I'm 42 years old."
+```
+
+Both of these functions accepts an optional `fallback` parameter, which is a string that will be used as a fallback
+value if a variable is not found in the given arguments. If no fallback is provided, an exception will be thrown.
+
+```kt
+val string = "Hello I'm {name}, and I'm {age} years old.".interpolate(
+    "name" to "John",
+    fallback = "unknown"
+)
+print(string) // prints "Hello I'm John, and I'm unknown years old."
+```
+
+The second variant comes in two flavors: `interpolate(vararg args: Any)` and `interpolate(args: List<Any>)`. They take
+a vararg or a list of values, and will replace the variables in the string using their index in the list. The variable
+names in the string must be integers, and in the range of the list. If a variable name is not an integer, or is out of
+bounds, an exception will be thrown.
+
+```kt
+val string = "'Hello I'm {0}, and I'm {1} years old.' said {0}.".interpolate(
+    "John",
+    42,
+)
+print(string) // prints "'Hello I'm John, and I'm 42 years old.' said John"
+```
+
+### Custom Variable Resolvers
+
+The previously presented functions are all using under the hood the default interpolation function
+`interpolate(VariableResolver)`. This function takes a `VariableResolver` as a parameter, which is a functional
+interface representing a function that takes a variable name and returns its value.
+
+Here is a dumb but simple example of how to use it:
+
+```kt
+val myResolver = VariableResolver { name -> name.reversed() }
```

Comment:
```suggestion
val myResolver = VariableResolver { variableName -> variableName.reversed() }
```

just to make sure to avoid confusion with the variable "name" in the string below